### PR TITLE
Reduce Route memory utilization by sharing the most common prefix path matchers

### PR DESCRIPTION
**Pull request by yurykats**
Reduce Route memory utilization by sharing the most common prefix path matchers

An analysis of a massively multi-tenant Envoy deployment shows that 75% of path matchers have either "" or "/" as a prefix. This PR creates a shared instance for both, so that all the routes can re-use the same trivial matcher instead of creating new identical ones.

This is a step towards more memory efficient config data structures. Issue #24154.

Signed-off-by: Yury Kats <ykats@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/common/matchers.cc [hunk 1]
```diff
@@ -130,12 +130,26 @@ PathMatcherConstSharedPtr PathMatcher::createExact(const std::string& exact, boo
...
+namespace {
+PathMatcherConstSharedPtr createPrefixPathMatcher(const std::string& prefix, bool ignore_case) {
   envoy::type::matcher::v3::StringMatcher matcher;
   matcher.set_prefix(prefix);
   matcher.set_ignore_case(ignore_case);
   return std::make_shared<const PathMatcher>(matcher);
 }
+static const PathMatcherConstSharedPtr emptyPrefixPathMatcher = createPrefixPathMatcher("", false);
+static const PathMatcherConstSharedPtr slashPrefixPathMatcher = createPrefixPathMatcher("/", false);
+} // namespace
```
Comment (pradeepcrao) [inline source/common/common/matchers.cc:142]: nit: blank line before `} // namespace`

File: test/common/common/matchers_test.cc [hunk 2]
```diff
@@ -432,6 +432,30 @@ TEST(PathMatcher, MatchPrefixPathIgnoreCase) {
...
+  const auto matcher3 = Envoy::Matchers::PathMatcher::createPrefix("/", true);
+  EXPECT_TRUE(matcher3->match("/bla"));
+
+  EXPECT_EQ(matcher1, matcher2);
+  EXPECT_EQ(matcher1, matcher3);
+}
+
+TEST(PathMatcher, EmptyPrefixMatcherShared) {
+  const auto matcher1 = Envoy::Matchers::PathMatcher::createPrefix("", false);
+  EXPECT_TRUE(matcher1->match("/bla"));
```
Comment (pradeepcrao) [inline test/common/common/matchers_test.cc:449]: nit: can the match input not begin with a `/`?

**Review by pradeepcrao — COMMENTED** [review]
LGTM.

File: source/common/common/matchers.cc [hunk 3]
```diff
@@ -130,13 +130,30 @@ PathMatcherConstSharedPtr PathMatcher::createExact(const std::string& exact, boo
...
   envoy::type::matcher::v3::StringMatcher matcher;
   matcher.set_prefix(prefix);
   matcher.set_ignore_case(ignore_case);
   return std::make_shared<const PathMatcher>(matcher);
 }
 
+} // namespace
+
+PathMatcherConstSharedPtr PathMatcher::createPrefix(const std::string& prefix, bool ignore_case) {
+  if (prefix.empty()) {
```
Comment (adisuissa) [inline source/common/common/matchers.cc:144]: nit: add comments explaining why this is added this way (preventing future "optimizations" to remove redundant code).

File: source/common/common/matchers.cc [hunk 4]
```diff
@@ -130,13 +130,30 @@ PathMatcherConstSharedPtr PathMatcher::createExact(const std::string& exact, boo
...
+    static const PathMatcherConstSharedPtr emptyPrefixPathMatcher =
+        createPrefixPathMatcher("", false);
+    return emptyPrefixPathMatcher;
```
Comment (adisuissa) [inline source/common/common/matchers.cc:147]: I guess that to use `CONSTRUCT_ON_FIRST_USE`, one will need to pass to it a function `createEmptyPrefixPathMatcher()` that essentially does `return createPrefixPathMatcher("", false);`.

I guess it complicates the readability of this code.

**Review by adisuissa — COMMENTED** [review]
Thanks, LGTM, modulo small nit.
I guess that these are the only 2 "well-used" prefix matchers, or otherwise a solution that uses a cache of prefix path-matchers would be better suited.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Review by zuercher — APPROVED** [review]

**Pull request merged**